### PR TITLE
Fix/issue 29899

### PR DESCRIPTION
**Pull request by Tony-MK**
### Details

https://github.com/Expensify/App/issues/29899
### Fixed Issues

$ https://github.com/Expensify/App/issues/29899
PROPOSAL: https://github.com/Expensify/App/issues/29899#issuecomment-1769823879

### Tests

1. Create a distance request, if none exists
2. Go Offline
3. Change the waypoints in the existing distance request 
4. Verify that the amount is replaced with TBD
5. Verify that the distance in the merchant string is replaced with TBD

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

[Android - Native.webm](https://github.com/Expensify/App/assets/22982173/e3c4af62-aeea-47d3-9387-a34a6a18e803)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

[Android - Chrome.webm](https://github.com/Expensify/App/assets/22982173/febfb3c5-9ead-4441-9ec4-592a5f5b8761)

</details>

<details>
<summary>iOS: Native</summary>

[iOS - Native.webm](https://github.com/Expensify/App/assets/22982173/862f08b9-351f-4f4f-b8c9-f955f8e204de)

</details>

<details>
<summary>iOS: mWeb Safari</summary>

[iOS - Safari.webm](https://github.com/Expensify/App/assets/22982173/3ff53ae8-7bc6-4d72-aa57-4b1f1d76b386)

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/22982173/0c1677b9-1fcc-45df-a586-c080859ba388

[macOS - Safari.webm](https://github.com/Expensify/App/assets/22982173/2c0f69aa-c8fe-41bf-aba7-889728da9d1e)

</details>

<details>
<summary>MacOS: Desktop</summary>

[macOS - Desktop.webm](https://github.com/Expensify/App/assets/22982173/a5945764-553c-4232-be93-a548ff3253de)

</details>

### CODE REVIEW

File: src/components/ReportActionItem/ReportPreview.js [hunk 1]
```diff
@@ -129,17 +129,24 @@ function ReportPreview(props) {
...
-    const previewSubtitle = hasOnlyOneReceiptRequest
-        ? TransactionUtils.getMerchant(transactionsWithReceipts[0])
-        : props.translate('iou.requestCount', {
-              count: numberOfRequests,
-              scanningReceipts: numberOfScanningReceipts,
-          });
+    let formattedMerchant = numberOfRequests === 1 && hasReceipts ? TransactionUtils.getMerchant(transactionsWithReceipts[0]) : null;
+    const hasPendingWaypoints = formattedMerchant && hasOnlyDistanceRequests && _.every(transactionsWithReceipts, (transaction) => lodashGet(transaction, 'pendingFields.waypoints', null));
+    if (hasPendingWaypoints) {
+        formattedMerchant = formattedMerchant.replace(/.+?(?=\s)/, props.translate('common.tbd'));
```
Comment (ArekChr) [inline src/components/ReportActionItem/ReportPreview.js:135]: I see this regex is used in 3 places, it should be moved to consts

File: src/components/ReportActionItem/ReportPreview.js [hunk 2]
```diff
@@ -129,17 +129,24 @@ function ReportPreview(props) {
...
+        formattedMerchant ||
+        props.translate('iou.requestCount', {
+            count: numberOfRequests,
+            scanningReceipts: numberOfScanningReceipts,
+        });
 
     const shouldShowSubmitButton = isReportDraft && reimbursableSpend !== 0;
 
     const getDisplayAmount = () => {
+        if (hasPendingWaypoints) {
```
Comment (ArekChr) [inline src/components/ReportActionItem/ReportPreview.js:147]: Two conditions could be combined
```suggestion
        if (hasPendingWaypoints || hasOnlyDistanceRequests) {
```

File: src/libs/ReceiptUtils.ts [hunk 3]
```diff
@@ -36,17 +36,19 @@ function getThumbnailAndImageURIs(transaction: Transaction, receiptPath: string
 
     const hasEReceipt = transaction?.hasEReceipt;
 
-    if (hasEReceipt) {
-        return {thumbnail: null, image: ROUTES.ERECEIPT.getRoute(transaction.transactionID), transaction};
-    }
+    if (!Object.hasOwn(transaction?.pendingFields ?? {}, 'waypoints')) {
```
Comment (ArekChr) [inline src/libs/ReceiptUtils.ts:39]: I think this could be simplified
```suggestion
    if (!transaction?.pendingFields?.waypoints) {
```

Comment (Tony-MK) [reply src/libs/ReceiptUtils.ts]: @ArekChr This line is making the type check in typescript fail. How should we change the code so the checks pass? 😄 Should we revert the line?

Comment (ArekChr) [reply src/libs/ReceiptUtils.ts]: I see now, let's leave it as it is

Comment (ArekChr) [reply src/libs/ReceiptUtils.ts]: @Tony-MK, I mean, let's leave it as it was implemented in the previous condition. :D

**Comment by ArekChr** [comment]
@Tony-MK TypeScript check fails, please fix the issues

**Comment by Tony-MK** [comment]
> @Tony-MK TypeScript check fails, please fix the issues and update branch with main

Fixed 😄

**Comment by Tony-MK** [comment]
Hey @ArekChr, Sorry for the delays. If you don't mind, there are two things I'm a bit curious about.

1) There are some components that are not consistent with distance requests that have recently been created offline and are TBD. Such as the money request header and total sum are supposed to show TBD. Also, the opacity of the amount and date fields are dimmed. Should I make it consistent with distance requests made offline?

2) I believe there is a bug with the `ReportPreview` component. When a user changes the waypoints offline, two or more map images will be displayed. One will show the pending map due to the changes we have made, and the rest will show the original map or another map. It seems to have two or more transactions when the waypoints change for a singular distance request. This bug currently occurs in staging and production. I can't seem to figure out which appropriate logic to use so that the `ReportPreview` component shows TBD and the correct number of maps when every request in the report is a distance request and all are TBD. What should we do about it?

**Comment by ArekChr** [comment]
> Should I make it consistent with distance requests made offline?

I think we should stick to one topic in this PR and talk about other adjustments in a different issue to keep things clear and organized.

File: src/CONST.ts [hunk 4]
```diff
@@ -803,7 +803,7 @@ const CONST = {
     },
 
     // at least 8 characters, 1 capital letter, 1 lowercase number, 1 number
-    PASSWORD_COMPLEXITY_REGEX_STRING: '^(?=.*[A-Z])(?=.*[0-9])(?=.*[a-z]).{8,}$',
+    PASSWORD_COMPLEXITY__STRING: '^(?=.*[A-Z])(?=.*[0-9])(?=.*[a-z]).{8,}$',
```
Comment (ArekChr) [inline src/CONST.ts:806]: I wanted to ask why this regex has been renamed, but it seems it is not used anywhere in the codebase. Could we remove it in this PR @arosiclair?

**Comment by ArekChr** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

File: src/CONST.ts (hunk 4 above)
Comment (arosiclair) [reply src/CONST.ts:806]: Yeah please revert this change @Tony-MK it doesn't seem necessary

Comment (Tony-MK) [reply src/CONST.ts:806]: Sorry for that; it is a mistake. Let me revert it.

**Comment by Tony-MK** [comment]
Apologies @arosiclair and @ArekChr for my lack of punctuality during the week. I have been handling a family emergency since Monday. However, I have made the suggested changes and updated it with the main branch.

Could we please get the PR reviewed so it can be delayed ASAP?

Thank you and have a fantastic day 😄.

**Comment by Tony-MK** [comment]
Apologies @arosiclair and @ArekChr for my lack of punctuality during the week. I have been handling a family emergency since Monday. However, I have made the suggested changes and updated it with the main branch.

Could we please get the PR reviewed so it can be delayed ASAP?

Thank you and have a fantastic day 😄.

File: src/CONST.ts (hunk 4 above)
Comment (ArekChr) [reply src/CONST.ts]: @arosiclair, I mean this regex is not used anywhere. Can we remove it in this PR to make the code cleaner?

**Comment by Tony-MK** [comment]
@ArekChr, It seems that I cannot reproduce the regression. Can you kindly explain how you reproduced it?

**Comment by ArekChr** [comment]
> @ArekChr, It seems that I cannot reproduce the regression. Can you kindly explain how you reproduced it? Could it related to the second issue in this https://github.com/Expensify/App/pull/30232#issuecomment-1781101722?

Yes, I thought this is different case but this is related with this issue, we should fix it in this PR

**Comment by Tony-MK** [comment]
> > @ArekChr, It seems that I cannot reproduce the regression. Can you kindly explain how you reproduced it? Could it related to the second issue in this [#30232 (comment)](https://github.com/Expensify/App/pull/30232#issuecomment-1781101722)?
> 
> Yes, I thought this is different case but this is related with this issue, we should fix it in this PR

Okay, then could you kindly explain to me how you reproduced it? I have tried to reproduce it numerous times but have failed.

**Comment by ArekChr** [comment]
1. Switch to offline mode.
2. Create and send a distance request to any workspace.
3. Navigate to the workspace where you sent the previously created request.
4. Observe the preview; the amount should display as "TBD."
5. Return to online mode and verify that the result remains unchanged.

**Comment by Tony-MK** [comment]
@ArekChr, Can you check again if the regression is still occurring? It seems to have been fixed when I updated the branch with main.

**Comment by Tony-MK** [comment]
Hey @ArekChr and @arosiclair, I have an important proposal to share with you about debugging the `ReportPreview` component. Accordin‌g to this [comment](https://github.com/Expensify/App/pull/30232#issuecomment-1782742344) I posted earlier, there are duplicate transactions, with the same `transactionId`, retrieved from the `ReportUtils.getTransactionsWithReceipts` function.

https://github.com/Expensify/App/blob/3db5ff4a1924156c748291d25eb9d774d82f879e/src/components/ReportActionItem/ReportPreview.js#L123

When a transaction's waypoints get updated, another transaction with the same transaction ID is stored in `allTransactions`. This is the main reason why you will notice one or more extra images maps appear in the `ReportPreview`. When the user goes back online, the older transaction(s) will get deleted and numbers of image maps will change and become the number of distance transactions.

Therefore, I kindly ask if it is possible of me to solve this bug. It's also affecting our ability to determine whether to show 'TBD' in the `ReportPreview` component when the `hasOnlyDistanceRequests` variable is `true` and all of the distance requests are waiting on the waypoints to be updated. Hence, my solution is after calling `ReportUtils.getTransactionsWithReceipts` function, we filter the latest transactions for every transaction ID and store the transactions in `transactionsWithReceipts`. I believe is a good method of insuring the we display the correct number of maps and TBD appropriately. I would greatly appreciate your thoughts on my proposal. Thank you 😄

**Comment by arosiclair** [comment]
> When a request's waypoints gets updated, another transaction with the same transaction ID is stored in allTransactions.

Can you find out why that's happening? This does not sound like correct behavior and it sounds like if we fix this, then your TBD problem should also be fixed.

**Comment by Tony-MK** [comment]
Hey @arosiclair and @ArekChr, I have been investigating the root cause of the second problem I mentioned in https://github.com/Expensify/App/pull/30232#issuecomment-1781101722 and elaborated in https://github.com/Expensify/App/pull/30232#issuecomment-1789872974.

When a distance request is edited, a backup copy of the original transaction. If the user doesn't save the changes to the transaction, the transaction data is restored from the backup copy. Read the code snippet below on the `useEffect` hook used in `EditRequestDistancePage` to further understand the significance of creating a backup copy of a transaction.

https://github.com/Expensify/App/blob/7bd51eba84f6708409dfd6e6886bf968eff3358f/src/pages/EditRequestDistancePage.js#L69-L87

The moment the server responds successfully, the backup transaction is deleted and is handled in the code snippet below.

https://github.com/Expensify/App/blob/7bd51eba84f6708409dfd6e6886bf968eff3358f/src/libs/actions/IOU.js#L785-L792

Therefore, the root cause of the problem is that the [`ReportUtils.getTransactionsWithReceipts`](https://github.com/Expensify/App/blob/7bd51eba84f6708409dfd6e6886bf968eff3358f/src/libs/ReportUtils.js#L1673) function, which is used for fetching transactions to display in the `ReportPreview` component, does not filter out the backup copies of the original transactions fetched from the [`TransactionUtils.getAllReportTransactions`](https://github.com/Expensify/App/blob/7bd51eba84f6708409dfd6e6886bf968eff3358f/src/libs/TransactionUtils.ts#L387) function. The functions return both the edited transaction and, if present, the backup of the original transaction.

https://github.com/Expensify/App/blob/7bd51eba84f6708409dfd6e6886bf968eff3358f/src/libs/ReportUtils.js#L1710-L1713

https://github.com/Expensify/App/blob/7bd51eba84f6708409dfd6e6886bf968eff3358f/src/libs/TransactionUtils.ts#L387-L393

In conclusion, if we filter out the backup copies of the original transactions, it would solve the problem mentioned in the first paragraph.

**Comment by arosiclair** [comment]
Very good breakdown @Tony-MK thanks! 

> In conclusion, if we filter out the backup copies of the original transactions, it would solve the problem mentioned in the first paragraph.

I agree with this. We can add an `isBackup` flag to the backup transaction when we create it and filter on that, though we would have to make sure we clear the flag before restoring the backup. What do you think?

I also do get the feeling that including the backup transaction in the normal transactions collection is probably going to cause more problems than it solves (like this one) so maybe we should consider moving backups to their own key. cc @tgolen since I think you designed this part

**Comment by tgolen** [comment]
I really like the idea of moving it to a separate collection. I wish I would have thought of that in the first place. Can you make that change? It might be nice to create an onyx migration to clean out backup transactions from the main collection.

**Comment by Tony-MK** [comment]
> I also do get the feeling that including the backup transaction in the normal transactions collection is probably going to cause more problems than it solves (like this one) so maybe we should consider moving backups to their own key. cc @tgolen since I think you designed this part

Thank you, @arosiclair and @tgolen for your feedback. I believe we are moving forward with creating an onyx migration.

> Can you make that change?

@tgolen, could you clarify to whom you were referring to make the change? Frankly, I wouldn't mind making the change.

**Comment by tgolen** [comment]
Yeah, @Tony-MK I think you can go ahead and make that change. I actually took the lead with that on this PR I've been writing to refactor the request creation flow.

Can you make [this same collection](https://github.com/Expensify/App/pull/28618/files#diff-facfc70b32eba19ccaeeadd16b51faf6dbf2ceb919373565e7322c86047886cdR266) and use that for the backup transactions?

**Comment by Tony-MK** [comment]
> Can you make [this same collection](https://github.com/Expensify/App/pull/28618/files#diff-facfc70b32eba19ccaeeadd16b51faf6dbf2ceb919373565e7322c86047886cdR266) and use that for the backup transactions?

Sure, @tgolen, I can certainly do that.

But before we proceed, I have a final question. Since this PR was originally intended for a different purpose than our discussion, should I create a new PR to ensure everything remains organized?

**Comment by tgolen** [comment]
I'd be fine with doing that in a different PR, sure.

**Comment by ArekChr** [comment]
@Tony-MK Please update PR with latest main

**Comment by ArekChr** [comment]
Should we display "TBD" in online mode optimistically, or is it better to keep it as is and update the amount only after the API returns successful data? @tgolen @arosiclair thoughts? 

https://github.com/Expensify/App/assets/24796318/a3a6d149-e570-4560-ba77-49231fe02e18

**Comment by Tony-MK** [comment]
> @Tony-MK Please update PR with latest main

Definitely and welcome back, @ArekChr

**Review by ArekChr — APPROVED** [review]

**Review by arosiclair — APPROVED** [review]
